### DISCUSSION TITLE
Best practice on testing wheels in forks

### DISCUSSION THREAD

**Discussion by abitrolly** [pypa/cibuildwheel#1164] [Q&A]
I often submit wheels builders to projects I do not own, and the code that uploads wheels is never tested, because it is going to fail as I don't have upload access. And even if I could, the repo to upload should be https://test.pypi.org/ with an ability to purge package and upload again, to avoid unnecessary bumping package versions.

Before inventing my own set of hacks & patches (tm) I want to ask if there is perhaps a best practices guide that covers this.g

**Comment by henryiii** [comment]
I've seen this done occasionally by owners. You can set up two steps, one that runs a test PyPI upload and the other that does the real upload, using `if:` to run one over the other. Though personally I've usually just added the test PyPI URL, tried it out, then removed the URL, rather than trying to keep uploading there. I think test PyPI works like the real PyPI and you can't purge it manually. You also can't get a project someone else owns there, so taking it from the real owners (if they haven't claimed it yet) is likely not ideal, and would likely be frowned on.

**Reply by henryiii** [reply]
Ahh, here's one that does the Test PyPI toggle: https://github.com/scikit-hep/pyhf/blob/master/.github/workflows/publish-package.yml (pure Python, but the publish step is the same as if you were pulling from a build wheel step).

**Reply by abitrolly** [reply]
So there should be a way to prefix package name for uploading to avoid conflicts.

**Reply by henryiii** [reply]
I think you'd be introducing more breaking points than what you are trying to test; especially with PEP 621, where the name can't be dynamic; tools like hatchling expect the package name to match the import name. So you'd add a little complexity now and potentially much more later if you move to static config to test uploading. PyPI is a bit more strict than `twine check`, and mistakes in upload can happen, but not sure it's worth it.

(I'm not sure if just renaming the file would pass a PyPI upload, I'd guess not since the name is in the metadata)

**Reply by abitrolly** [reply]
Given all the above, I conclude there is no way to test wheels in forks with TestPyPI if you don't own the package name.

**Reply by henryiii** [reply]
I think the best practice is to test the wheels made by the job and hope the upload works. If you’ve done a few (and especially of the metadata is already right due to SDist uploads working), then it’s pretty safe.

Or you can ask the package owner if you can take or be added to the TestPyPI project; I’d think most would be happy to allow anyone to play with TestPyPI if they are developing wheels. For the reupload version issue, you’d do dev or alpha releases, something you can easily increment and not interfere with the actual planned releases (which won’t go to TestPyPI).

You could also host a wheelhouse yourself. :)

**Reply by henryiii** [reply]
In terms of the issue though, not sure there’s a good best practice that works in all cases.

**Reply by abitrolly** [reply]
> You could also host a wheelhouse yourself. :)

Actually that sounds like an option.

**Reply by abitrolly** [reply]
https://hub.docker.com/r/pypa/warehouse don't seem to be huge, but they are dated. Need to run some tests.
